module/VuFind/src/VuFind/View/Helper/Root/Citation.php

```diff
@@ -262,16 +264,181 @@ function (string $value) use ($callables): string {
+
+        $processor = new CiteProc(StyleSheet::loadStyleSheet($format), $locale);
+
+        // DEBUG
+
         // Construct method name for requested format:
-        $method = 'getCitation' . $format;
+        $formatFirst = explode('-', $format, 2)[0];
+        $methodKey = ($formatFirst == 'modern') ? 'MLA' : ucfirst($formatFirst);
+
+        $method = 'getCitation' . $methodKey;
 
         // Avoid calls to inappropriate/missing methods:
         if (!empty($format) && method_exists($this, $method)) {
-            return $this->$method();
+            return '<ul>' .
+                '<li><kbd>Old</kbd> ' . $this->$method() . '</li>' .
+                '<li><kbd>New</kbd> ' . $processor->render(json_decode($data), 'bibliography') . '</li>' .
+            '</ul>';
+        }
+
+        return $processor->render(json_decode($data), 'bibliography');
+    }
+
+    protected function trimPunctuation($text)
+    {
+        return trim($text, " \n\r\t\v\0,:;/");
+    }
+
+    protected function removeDateRange($name)
+    {
+        return preg_replace('/\d+[ ]*\-[ ]*\d*/', '', $name);
+    }
+
+    protected function nameToGivenFamily($name)
+    {
+        if (strpos($name, ', ') !== false) {
+            [$family, $given] = explode(', ', $this->removeDateRange($name));
+
+            return [
+                'given' => $this->trimPunctuation($given),
+                'family' => $this->trimPunctuation($family),
+            ];
+        }
+
+        $parts = explode(' ', $this->removeDateRange($name));
+
+        $family = array_pop($parts);
+        $given = implode(' ', $parts);
+
+        return [
+            'given' => $this->trimPunctuation($given),
+            'family' => $this->trimPunctuation($family),
+        ];
+    }
+
+    protected function addIfNotEmpty(&$item, $pairs)
+    {
+        foreach ($pairs as $key => $value) {
+            if (!empty($value)) {
+                $item[$key] = $this->trimPunctuation(((array) $value)[0]);
+            }
+        }
+    }
+
+    /**
+     * Map data about the current record to the CSL JSON schema defined here:
+     * https://github.com/citation-style-language/schema/blob/master/csl-data.json
+     *
+     * @return string
+     */
+    public function getDataCSL()
+    {
+        // id
+        $item = ['id' => $this->driver->getUniqueID()];
+
+        // type
+        switch ($this->driver->getFormats()[0]) {
+        case 'Thesis':
+            $item['type'] = 'thesis';
+            break;
+        case 'Video':
+            $item['type'] = 'motion_picture';
+            break;
+        case 'Score':
+            $item['type'] = 'musical_score';
+            break;
+        case 'Map':
+            $item['type'] = 'map';
+            break;
+        case 'Book':
+        default:
+            $item['type'] = 'book';
+        }
+
+        // title
+        $this->addIfNotEmpty(
+            $item,
+            [
+                'title' => $this->driver->tryMethod('getTitle'),
```

Comment:
Since CSL doesn't have a subtitle field, some implementations add the subtitle here, but i'm not sure if that's correct in 100% of the cases. What do you think about it?

Reply:
The `getTitle` method should already include the subtitle -- it returns the entire title, whereas `getShortTitle` + `getSubtitle` allow you to retrieve the individual parts.